KAFKA-2841: safe group metadata cache loading/unloading

hachikuji opened the pull request.

hachikuji committed:
KAFKA-2841: safe group metadata cache loading/unloading

hachikuji committed:
address Ismael's review comments

guozhangwang commented on core/src/main/scala/kafka/coordinator/GroupCoordinator.scala:
```diff
@@ -473,12 +473,55 @@ class GroupCoordinator(val brokerId: Int,
     }
   }
 
+  private def onGroupUnloaded(groupEmigrating: Boolean)(group: GroupMetadata) {
```
Could we move these two functions into GroupMetadataManager to avoid function passing with separate arguments?

hachikuji replied:
I slightly favor keeping them here since that keeps all state transitions and response callback handling confined to GroupCoordinator, but I don't have a strong preference if that is easier to understand.

guozhangwang commented:
In an extreme case where a partition leader jumps from A to B and then back to A in a short time, coordinator A will get two LeaderAndISR requests, one for emigrating the groups out and one for loading the groups back in. Since they can be handled by different threads, handleGroupImmigration maybe triggered before handleGroupEmmigration, how do we handle this case?

EDIT: actually this should not happen, since this case is guarded by `replicaStateChangeLock` in replica manager.

guozhangwang commented:
This fix looks good to me overall but is a bit hard to understand, I am wondering if we could just move the callbacks into metadata manager?

guozhangwang commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -409,21 +403,37 @@ class GroupMetadataManager(val brokerId: Int,
+
+              removedGroups.foreach { groupId =>
+                val group = groupsCache.get(groupId)
+                if (group != null) {
```
We can at least add a warning here about this case that should not happening, or even throw an exception to fail the coordinator in this case. If we do the periodic group purging later then we can remove the warning / exception thrown.

hachikuji committed:
address review comments; start heartbeat when group is loaded

hachikuji committed:
ensure proper order of group immigration/emigration

hachikuji commented:
@guozhangwang @junrao Ping for review. To solve the group load/unload ordering problem, I did the following:
1. Refactored GroupMetadataManager to use a single-thread scheduler.
2. Moved the work of removeGroupsForPartition into the scheduler.
3. Passed through the controller/leader epochs to ensure proper ordering of loading/unloading.

The use of the single-thread scheduler seems unfortunate, but we already had a global lock for loading partitions, so it's not too bad. In fact, it's probably better not to tie up additional threads if everything has to be synchronized anyway.

junrao commented:
Basically, if a leaderAndIsrRequest with an older leaderEpoch comes in, ReplicaManager will reject it and not include the partition in the updatededLeader field in the return value of becomeLeaderOrFollower(). So, we can expect calls to do groupImmigration is always in increasing leaderEpoch.

hachikuji commented:
@junrao I think the question is whether a reordering can occur between the time becomeLeaderOrFollower() exits and the time the load/unload tasks are scheduled and executed. As long as we can guarantee that calls to schedule() are called in the correct order, it looks like we can depend on the scheduler to ensure the expected order of execution. Since the server does not permit concurrent handling of requests from the same controller, it seems like that guarantee holds, so I think it might be sufficient to check only the controller epoch.

hachikuji committed:
restructure group loading/unloading to be called under replicate state lock

hachikuji commented:
@junrao Ready for another look.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -112,23 +113,23 @@ class KafkaApis(val requestChannel: RequestChannel,
-        partition.leaderReplicaIdOpt.foreach { leaderReplica =>
-          if (partition.topic == GroupCoordinator.GroupMetadataTopicName &&
-              leaderReplica == brokerId)
+      def onLeadershipChange(updatedLeaders: Iterable[Partition], updatedFollowers: Iterable[Partition]): Unit = {
```
We can just do the following and get rid of : Unit =
def onLeadershipChange(updatedLeaders: Iterable[Partition], updatedFollowers: Iterable[Partition]) {

junrao commented:
Thanks for the patch. Just a minor comment. Otherwise, LGTM.

hachikuji committed:
remove unneeded return type

hachikuji commented:
@junrao Fixed the callback return type.

guozhangwang commented:
Thanks, LGTM.

guozhangwang commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -77,8 +76,12 @@ class GroupMetadataManager(val brokerId: Int,
   private val groupMetadataTopicPartitionCount = getOffsetsTopicPartitionCount
 
+  /* Single-thread scheduler to handling offset/group metadata cache loading and unloading */
+  private val scheduler = new KafkaScheduler(threads = 1, threadNamePrefix = "group-metadata-manager-")
```
May be comment on the motivation why we need it to be single-threaded.

junrao commented:
Thanks for the patch. LGTM
